File: src/memory/builder/release_intent.py
"""Delivery Story release intent without release authorization."""

from __future__ import annotations

import re
from dataclasses import dataclass

from memory.builder.delivery_cursor import (
    BuilderDeliveryCursor,
    get_delivery_cursor,
    set_delivery_cursor,
)
from memory.builder.surface_protocol import wrap_ariad_surface
from memory.storage.store import Store

RELEASE_INTENTS = ("planned", "none", "undecided")


@dataclass(frozen=True)
class ReleaseIntentReport:
    journey: str
    method: str
    delivery_story: str
    intent: str
    cursor: BuilderDeliveryCursor
    changed: bool


def set_release_intent(
    store: Store,
    *,
    journey: str,
    method: str,
    intent: str,
) -> ReleaseIntentReport:
    """Record informational release intent for the active Delivery Story boundary."""
    normalized = intent.strip().lower()
    if normalized not in RELEASE_INTENTS:
        raise ValueError("release intent must be planned, none, or undecided")
    cursor, delivery_story = _active_delivery_story(store, journey, method)
    changed = not (
        cursor.release_intent_delivery_story == delivery_story
        and cursor.release_intent == normalized
    )
    updated = _persist_release_intent(
        store,
        cursor,
        delivery_story=delivery_story,
        intent=normalized,
    )
    return ReleaseIntentReport(
        journey=journey,
        method=method,
        delivery_story=delivery_story,
        intent=normalized,
        cursor=updated,
        changed=changed,
    )


def inspect_release_intent(
    store: Store,
    *,
    journey: str,
    method: str,
) -> ReleaseIntentReport:
    """Inspect release intent distinctly from an explicit ``none`` state."""
    cursor, delivery_story = _active_delivery_story(store, journey, method)
    intent = (
        cursor.release_intent
        if cursor.release_intent_delivery_story == delivery_story and cursor.release_intent
        else "not_recorded"
    )
    return ReleaseIntentReport(
        journey=journey,
        method=method,
        delivery_story=delivery_story,
        intent=intent,
        cursor=cursor,
        changed=False,
    )


def delivery_story_code_for_item(item_code: str | None) -> str | None:
    """Return the authored Delivery Story ancestor for supported Ariad codes."""
    if not item_code:
        return None
    parts = tuple(part.strip() for part in item_code.split(".") if part.strip())
    for index, part in enumerate(parts):
        if re.fullmatch(r"DS-?\d+", part, flags=re.IGNORECASE):
            return ".".join(parts[: index + 1])
    return None


def render_release_intent_report(report: ReleaseIntentReport) -> str:
    action = "recorded" if report.changed else "inspected"
    meaning = {
        "planned": "This Delivery Story is expected to create a release boundary if completed coherently.",
        "none": "This Delivery Story is not expected to create a release boundary.",
        "undecided": "Release intent is intentionally unresolved and may be revisited later.",
        "not_recorded": "No release intent has been recorded for this Delivery Story.",
    }[report.intent]
    body = "\n".join(
        [
            "Delivery",
            "",
            "╭────────────────────────────────────────────────────────╮",
            "│        🚦  DELIVERY STORY RELEASE INTENT               │",
            "│                                                        │",
            _card_text("Delivery Story"),
            *_card_wrapped(report.delivery_story),
            "│                                                        │",
            _card_text("Intent"),
            *_card_wrapped(report.intent),
            "│                                                        │",
            _card_text("Meaning"),
            *_card_wrapped(meaning),
            "│                                                        │",
            _card_text("Action"),
            *_card_wrapped(action),
            "│                                                        │",
            _card_text("Authority boundary"),
            *_card_wrapped(
                "Release intent does not authorize commit, push, tag creation, stable promotion, release publication, or remote mutation."
            ),
            "╰────────────────────────────────────────────────────────╯",
        ]
    )
    return wrap_ariad_surface("release_intent", body + "\n")


def _active_delivery_story(
    store: Store, journey: str, method: str
) -> tuple[BuilderDeliveryCursor, str]:
    cursor = get_delivery_cursor(store, journey)
    if cursor is None:
        raise ValueError("delivery cursor is required before release intent")
    if cursor.method != method:
        raise ValueError("active Builder method does not match release intent method")
    delivery_story = delivery_story_code_for_item(cursor.active_item)
    if delivery_story is None:
        raise ValueError("active Delivery Story boundary is required for release intent")
    return cursor, delivery_story


def _persist_release_intent(
    store: Store,
    cursor: BuilderDeliveryCursor,
    *,
    delivery_story: str,
    intent: str,
) -> BuilderDeliveryCursor:
    return set_delivery_cursor(
        store,
        journey=cursor.journey,
        method=cursor.method,
        active_item=cursor.active_item,
        active_item_title=cursor.active_item_title,
        active_item_level=cursor.active_item_level,
        active_checkpoint=cursor.active_checkpoint,
        pending_confirmation=cursor.pending_confirmation,
        last_delivery_event=cursor.last_delivery_event,
        cadence_profile=cursor.cadence_profile,
        cadence_limits=cursor.cadence_limits,
        granularity_decision=cursor.granularity_decision,
        navigator_flow_unit=cursor.navigator_flow_unit,
        child_work_items=cursor.child_work_items,
        aggregate_checkpoint_status=cursor.aggregate_checkpoint_status,
        cursor_generation=cursor.cursor_generation,
        release_intent_delivery_story=delivery_story,
        release_intent=intent,
    )


def _card_text(text: str) -> str:
    width = 54
    return f"│ {text[:width]:<{width}} │"


def _card_wrapped(text: str) -> list[str]:
    return [_card_text(line) for line in _wrap_plain_text(text, width=54)]


def _wrap_plain_text(text: str, *, width: int) -> list[str]:
    words = text.split()
    lines: list[str] = []
    current = ""
    for word in words:
        candidate = f"{current} {word}".strip()
        if len(candidate) > width and current:
            lines.append(current)
            current = word
        else:
            current = candidate
    if current:
        lines.append(current)
    return lines or ["none"]

File: ts/src/builder/releaseIntent.ts
// CV22.DS7.US8 plateau 6 — Delivery Story release intent.
//
// Port of `src/memory/builder/release_intent.py`.
//
// Release intent is the only Builder state that belongs to an ANCESTOR rather than
// to the active item: it is stored as a pair — the intent and the Delivery Story
// code it was recorded against — so moving to a story under a different Delivery
// Story makes the inspect face report `not_recorded` instead of inheriting the
// previous decision. Two cursor fields, one meaning.
//
// The boundary itself is derived from the active item's CODE by
// `deliveryStoryCodeForItem` (already ported for the cursor), so a story whose code
// carries no `DS<n>` segment has nothing to record against and is refused.
//
// `not_recorded` is a RENDERED value, never a stored one: the cursor holds `null`,
// and the surface distinguishes "nothing was decided" from the explicit decision
// `none`. Collapsing the two would erase the difference between an open question
// and an answered one, which is the whole point of the `undecided` value existing
// beside them.

import type { Database, WritableDatabase } from "#db/database.ts";
import { pyStrip } from "#util/pythonText.ts";
import { cardText, cardWrapped } from "./card.ts";
import { deliveryStoryCodeForItem } from "./cursorTransitions.ts";
import {
  type BuilderDeliveryCursor,
  type CursorWriteDeps,
  getDeliveryCursor,
  setDeliveryCursor,
} from "./deliveryCursor.ts";
import { wrapAriadSurface } from "./surfaceProtocol.ts";

/** Python `RELEASE_INTENTS`. */
export const RELEASE_INTENTS = ["planned", "none", "undecided"] as const;

export interface ReleaseIntentReport {
  readonly journey: string;
  readonly method: string;
  readonly deliveryStory: string;
  readonly intent: string;
  readonly cursor: BuilderDeliveryCursor;
  /** False when the pair already said this, which the surface renders as `inspected`. */
  readonly changed: boolean;
}

/** Python `_active_delivery_story`: cursor, method match, then the DS ancestor. */
function activeDeliveryStory(
  db: Database,
  journey: string,
  method: string,
): { cursor: BuilderDeliveryCursor; deliveryStory: string } {
  const cursor = getDeliveryCursor(db, journey);
  if (cursor === null) {
    throw new Error("delivery cursor is required before release intent");
  }
  if (cursor.method !== method) {
    throw new Error("active Builder method does not match release intent method");
  }
  const deliveryStory = deliveryStoryCodeForItem(cursor.activeItem);
  if (deliveryStory === null) {
    throw new Error("active Delivery Story boundary is required for release intent");
  }
  return { cursor, deliveryStory };
}

/** Python `set_release_intent`. */
export function setReleaseIntent(
  db: WritableDatabase,
  options: { journey: string; method: string; intent: string },
  deps: CursorWriteDeps,
): ReleaseIntentReport {
  const normalized = pyStrip(options.intent).toLowerCase();
  if (!(RELEASE_INTENTS as readonly string[]).includes(normalized)) {
    throw new Error("release intent must be planned, none, or undecided");
  }
  // Value first, then the cursor: an unknown intent is refused even with no cursor.
  const { cursor, deliveryStory } = activeDeliveryStory(db, options.journey, options.method);
  const changed = !(
    cursor.releaseIntentDeliveryStory === deliveryStory && cursor.releaseIntent === normalized
  );
  const updated = setDeliveryCursor(
    db,
    {
      journey: cursor.journey,
      method: cursor.method,
      activeItem: cursor.activeItem,
      activeItemTitle: cursor.activeItemTitle,
      activeItemLevel: cursor.activeItemLevel,
      activeCheckpoint: cursor.activeCheckpoint,
      pendingConfirmation: cursor.pendingConfirmation,
      lastDeliveryEvent: cursor.lastDeliveryEvent,
      cadenceProfile: cursor.cadenceProfile,
      cadenceLimits: cursor.cadenceLimits,
      granularityDecision: cursor.granularityDecision,
      navigatorFlowUnit: cursor.navigatorFlowUnit,
      childWorkItems: cursor.childWorkItems,
      aggregateCheckpointStatus: cursor.aggregateCheckpointStatus,
      cursorGeneration: cursor.cursorGeneration,
      releaseIntentDeliveryStory: { kind: "set", value: deliveryStory },
      releaseIntent: { kind: "set", value: normalized },
    },
    deps,
  );
  return {
    journey: options.journey,
    method: options.method,
    deliveryStory,
    intent: normalized,
    cursor: updated,
    changed,
  };
}

/** Python `inspect_release_intent`. */
export function inspectReleaseIntent(
  db: Database,
  options: { journey: string; method: string },
): ReleaseIntentReport {
  const { cursor, deliveryStory } = activeDeliveryStory(db, options.journey, options.method);
  const recorded =
    cursor.releaseIntentDeliveryStory === deliveryStory && cursor.releaseIntent
      ? cursor.releaseIntent
      : "not_recorded";
  return {
    journey: options.journey,
    method: options.method,
    deliveryStory,
    intent: recorded,
    cursor,
    changed: false,
  };
}

const MEANING: Record<string, string> = {
  planned: "This Delivery Story is expected to create a release boundary if completed coherently.",
  none: "This Delivery Story is not expected to create a release boundary.",
  undecided: "Release intent is intentionally unresolved and may be revisited later.",
  not_recorded: "No release intent has been recorded for this Delivery Story.",
};

/** Python `render_release_intent_report`. */
export function renderReleaseIntentReport(report: ReleaseIntentReport): string {
  const meaning = MEANING[report.intent];
  if (meaning === undefined) {
    // Python indexes the dict directly, so an unknown intent is a KeyError there.
    throw new Error(`unknown release intent: ${report.intent}`);
  }
  const body = [
    "Delivery",
    "",
    "╭────────────────────────────────────────────────────────╮",
    "│        🚦  DELIVERY STORY RELEASE INTENT               │",
    "│                                                        │",
    cardText("Delivery Story"),
    ...cardWrapped(report.deliveryStory),
    "│                                                        │",
    cardText("Intent"),
    ...cardWrapped(report.intent),
    "│                                                        │",
    cardText("Meaning"),
    ...cardWrapped(meaning),
    "│                                                        │",
    cardText("Action"),
    ...cardWrapped(report.changed ? "recorded" : "inspected"),
    "│                                                        │",
    cardText("Authority boundary"),
    ...cardWrapped(
      "Release intent does not authorize commit, push, tag creation, stable promotion, release publication, or remote mutation.",
    ),
    "╰────────────────────────────────────────────────────────╯",
  ];
  return wrapAriadSurface("release_intent", `${body.join("\n")}\n`);
}
